### PR TITLE
Fix for issue 635: check whether `self.mask` exists before using it

### PR DESCRIPTION
**Pull request by keflavich**
Corrects an error where some dask cubes can have `None` masks, causing errors.

It's unclear whether this is the right fix; a more appropriate fix might be to
systematically ensure that mask is never None

### CODE REVIEW

**Comment by keflavich** [comment]
On a little further inspection, we do have use cases that check for `mask is None`, so catching cases where mask can't be `None` is the right approach.

**Comment by keflavich** [comment]
Issue #635 wasn't caught by tests because we don't have a varying-resolution CASA .image cube.

@astrofrog how did we add CASA cubes to tests?  I think adding one to `test_mask_bad_beams` is the regression test we need here.

**Comment by keflavich** [comment]
Something is still broken:

```
test_mask_bad_beams('/Users/adam/repos/spectral-cube/spectral_cube/tests/data/vda_beams.image', True)
Traceback (most recent call last):
  File "<ipython-input-13-ac5f40bd34dc>", line 1, in <module>
    test_mask_bad_beams('/Users/adam/repos/spectral-cube/spectral_cube/tests/data/vda_beams.image', True)
  File "/Users/adam/repos/spectral-cube/spectral_cube/tests/test_spectral_cube.py", line 1890, in test_mask_bad_beams
    assert np.all(masked_cube.mask.include()[:,0,0] == [False,True,True,False])
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/array/core.py", line 1372, in __bool__
    return bool(self.compute())
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/base.py", line 175, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/base.py", line 446, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/threaded.py", line 82, in get
    **kwargs
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/local.py", line 491, in get_async
    raise_exception(exc, tb)
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/compatibility.py", line 130, in reraise
    raise exc
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/local.py", line 233, in execute_task
    result = _execute_task(task, data)
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/core.py", line 119, in _execute_task
    return func(*args2)
  File "/Users/adam/miniconda3/envs/casa_py36/lib/python3.6/site-packages/dask/array/core.py", line 98, in getter
    c = a[b]
  File "/Users/adam/repos/spectral-cube/spectral_cube/io/casa_dask.py", line 80, in __getitem__
    offset=start, count=end - start)
TypeError: 'offset' is an invalid keyword argument for this function
```

**Comment by keflavich** [comment]
Should this be removed: https://github.com/radio-astro-tools/spectral-cube/pull/639/commits/67f22a07672846cf990a73bddfbb563e6b090e3c and instead have the .image just generate on the fly?

**Comment by keflavich** [comment]
There are real test failures here.  This needs more work.

**Comment by keflavich** [comment]
This PR is on hold until I get feedback from @astrofrog.  Most likely we should not be adding test data, but instead should generate it.

**Comment by astrofrog** [comment]
Could you use https://github.com/radio-astro-tools/spectral-cube/tree/master/spectral_cube/tests/data/nomask.image? or is the issue that you need variable beams?

File: spectral_cube/conftest.py [hunk 1]
```diff
@@ -282,6 +282,26 @@ def data_vda_beams(tmp_path):
     return tmp_path / 'vda_beams.fits'
 
 
+@pytest.fixture
+def data_vda_beams_image(tmp_path):
```
Comment (astrofrog) [inline spectral_cube/conftest.py:286]: Yes this should be sufficient and no need to then commit the .image directory

You can also make it even simpler by using one of the FITS fixtures and this: https://github.com/radio-astro-tools/spectral-cube/blob/master/spectral_cube/tests/test_casafuncs.py#L151

**Comment by keflavich** [comment]
This issue has bitten me again.

**Comment by keflavich** [comment]
this is a deeper hole than I like:
```
Traceback (most recent call last):
  File "/orange/adamginsburg/ALMA_IMF/reduction/analysis/line_quicklooks.py", line 113, in <module>
    stdspec = mcube.mad_std(axis=(1,2))#, how='slice')
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/dask_spectral_cube.py", line 115, in wrapper
    out = function(self, *args, **kwargs)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/dask_spectral_cube.py", line 66, in wrapper
    return function(self, *args, **kwargs)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/dask_spectral_cube.py", line 672, in mad_std
    **kwargs))
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/dask_spectral_cube.py", line 284, in _compute
    return array.compute(**self._scheduler_kwargs)
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/base.py", line 167, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/base.py", line 447, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/local.py", line 527, in get_sync
    return get_async(apply_sync, 1, dsk, keys, **kwargs)
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/local.py", line 471, in get_async
    fire_task()
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/local.py", line 466, in fire_task
    callback=queue.put,
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/local.py", line 516, in apply_sync
    res = func(*args, **kwds)
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/local.py", line 227, in execute_task
    result = pack_exception(e, dumps)
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/local.py", line 222, in execute_task
    result = _execute_task(task, data)
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/core.py", line 121, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/dask/array/core.py", line 98, in getter
    c = a[b]
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/dask_spectral_cube.py", line 195, in __getitem__
    return self._cube._mask._filled(data=self._cube._data,
AttributeError: 'Array' object has no attribute '_filled'
```

**Comment by keflavich** [comment]
How did the above happen?  I can't see a way in any of spectral-cube to end up with a dask array as a mask:
```python
In [1]: %debug
> /orange/adamginsburg/repos/spectral-cube/spectral_cube/dask_spectral_cube.py(195)__getitem__()
    193             return 0.
    194         else:
--> 195             return self._cube._mask._filled(data=self._cube._data,
    196                                             view=view,
    197                                             wcs=self._cube._wcs,

ipdb> self._cube._mask
dask.array<logical_and, shape=(30, 1920, 1920), dtype=bool, chunksize=(30, 960, 960), chunktype=numpy.ndarray>
```
but that's what we have

File: spectral_cube/conftest.py (hunk 1 above)
Comment (keflavich) [reply spectral_cube/conftest.py:286]: I removed the `.image` directory, now we just use this.

**Comment by keflavich** [comment]
Somehow py3.6 is reading multi-beam cubes not as a varyingresolution cube:
https://travis-ci.org/github/radio-astro-tools/spectral-cube/jobs/737242230#L418

**Comment by keflavich** [comment]
I can't reproduce travis's errors locally any more.

**Comment by keflavich** [comment]
`tox -e test` is skipping the dask tests: `../../spectral_cube/io/tests/test_casa_dask.py ssssssssssssss`
while `py.test` is running them:
`spectral_cube/io/tests/test_casa_dask.py ..............`

`python setup.py test` fails with a verbose error message about `tox`.  Is any of this supposed to happen?

**Comment by keflavich** [comment]
locally, I get CASA failures where the equality test between two dictionaries fails because one of them explicitly prints the dtypes

**Comment by keflavich** [comment]
```

spectral_cube/io/tests/test_casa_dask.py ..............                                                                                                                                                                                                                                [  0%]
spectral_cube/io/tests/test_casa_low_level_io.py FFFFFFFF.FFFFFFFFFFFFFFFFFFFFFF.                                                                                                                                                                                                      [  2%]
spectral_cube/io/tests/test_casa_wcs.py ..................................................                                                                                                                                                                                             [  5%]
spectral_cube/tests/test_analysis_functions.py .................                                                                                                                                                                                                                       [  6%]
spectral_cube/tests/test_casafuncs.py ....................                                                                                                                                                                                                                             [  7%]
spectral_cube/tests/test_cube_utils.py .......                                                                                                                                                                                                                                         [  8%]
spectral_cube/tests/test_dask.py .s...                                                                                                                                                                                                                                                 [  8%]
spectral_cube/tests/test_io.py ..................                                                                                                                                                                                                                                      [  9%]
spectral_cube/tests/test_masks.py .....................................................................................XX....                                                                                                                                                          [ 14%]
spectral_cube/tests/test_moments.py ............................................................................................................................................................................................                                                       [ 25%]
spectral_cube/tests/test_performance.py ...s..                                                                                                                                                                                                                                         [ 26%]
spectral_cube/tests/test_projection.py .......................x.....................x................................................                                                                                                                                                  [ 31%]
spectral_cube/tests/test_regrid.py ......ssss....................sss....s                                                                                                                                                                                                              [ 33%]
spectral_cube/tests/test_spectral_axis.py ....................................................                                                                                                                                                                                         [ 36%]
spectral_cube/tests/test_spectral_cube.py .........................................s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.s.................................................................................... [ 50%]
......................................................................................................................................s.....s...........sssss.........................................ssssss......................ss.................................................. [ 66%]
........................................................................ssssss......ssssssssssss......ssssssssssss......ssssssssssss......ssssssssssss......ssssss..........................................................................................ssssss......ssssssssssss.. [ 82%]
....ssssssssssss......ssssssssssss......ssssssssssss......ssssss.................................s.s..............xx...................................E.E...................................................                                                                          [ 94%]
spectral_cube/tests/test_stokes_spectral_cube.py ..........................................                                                                                                                                                                                            [ 97%]
spectral_cube/tests/test_subcubes.py ..ssssssssssssssssssss                                                                                                                                                                                                                            [ 98%]
spectral_cube/tests/test_visualization.py ....ssssssss                                                                                                                                                                                                                                 [ 99%]
spectral_cube/tests/test_wcs_utils.py ............                                                                                                                                                                                                                                     [100%]
```

**Comment by keflavich** [comment]
local test failure summary:
```
================================================================================================================================== short test summary info ===================================================================================================================================
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdminfo[shape0] - assert "{'*1': {'COL...dCellStMan'}}" == "{'*1': {'COL...dCellStMan'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdminfo[shape1] - assert "{'*1': {'COL...dCellStMan'}}" == "{'*1': {'COL...dCellStMan'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdminfo[shape2] - assert "{'*1': {'COL...dCellStMan'}}" == "{'*1': {'COL...dCellStMan'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdminfo[shape3] - assert "{'*1': {'COL...dCellStMan'}}" == "{'*1': {'COL...dCellStMan'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdminfo[shape4] - assert "{'*1': {'COL...dCellStMan'}}" == "{'*1': {'COL...dCellStMan'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdminfo[shape5] - assert "{'*1': {'COL...dCellStMan'}}" == "{'*1': {'COL...dCellStMan'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdminfo[shape6] - assert "{'*1': {'COL...dCellStMan'}}" == "{'*1': {'COL...dCellStMan'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdminfo[shape7] - assert "{'*1': {'COL...dCellStMan'}}" == "{'*1': {'COL...dCellStMan'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_advs] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_dvsa] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_vsad] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_sadv] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_sdav] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_sdav_beams] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_advs_nobeam] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_adv] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_adv_jybeam_upper] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_adv_jybeam_lower] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_adv_jybeam_whitespace] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_adv_beams] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_vad] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_vda] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_vda_jybeam_upper] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_vda_jybeam_lower] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_vda_jybeam_whitespace] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_vda_beams] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_255] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_255_delta] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_getdesc[data_522_delta] - assert "{'_define_hy...e': 'float'}}" == "{'_define_hy...e': 'float'}}"
FAILED spectral_cube/io/tests/test_casa_low_level_io.py::test_generic_table_read - assert "{'_define_hy...': 'double'}}" == "{'_define_hy...': 'double'}}"
ERROR spectral_cube/tests/test_spectral_cube.py::test_mask_bad_beams[False-data_vda_beams_image] - RuntimeError: Exception: Error Illegal Quantum record in QuantumHolder::fromRecord
ERROR spectral_cube/tests/test_spectral_cube.py::test_mask_bad_beams[True-data_vda_beams_image] - RuntimeError: Exception: Error Illegal Quantum record in QuantumHolder::fromRecord
```

**Comment by astrofrog** [comment]
@keflavich - can you rebase so we can see if there are still failures? (now that https://github.com/radio-astro-tools/spectral-cube/pull/684 is merged)

**Comment by keflavich** [comment]
Passed locally!

**Review by astrofrog — APPROVED** [review]
This seems fine, feel free to ignore the coverage error although it does indicate some of the code here is not being tested?

**Comment by keflavich** [comment]
Yeah I think I didn't get the coverage to 100%, but I should have increased it...

**Comment by keflavich** [comment]
Some of the new test infrastructure is getting counted against the diff.  I'm calling this good.

**Pull request merged**